### PR TITLE
[WIP] Issue/gl fixes

### LINKED ISSUE #174: Broken display on windows 10
I upgraded to windows 10 recently and now getting this after launching:
image

I think this is a fairly recent build of performous (less then year).

### PR DESCRIPTION
**Pull request by Lord-Kamina**
### What does this PR do?

- Does away with most glmath.hh in favor of using glm, which is an actively maintained GL Math library.
- Fixes the way shaders are compiled and linked so the geometry shader is now actually optional, instead of having a hacky pass-through mode that always runs.
- Fixed some structs and other issues in the shaders, rewrote the geometry shader (fixes #174!)
- Rewrote ShaderMap in order to use STL containers and unique_ptr instead of boost:ptr_map.

Shit was broken... Like, really.

### Closes Issue(s)
* closes #174 

### Motivation
Seriously?

### More

- [ ] Added/updated documentation

### Additional Notes

Note: This PR is WIP because,

a) Currently it's just a duplicate of the win-master branch; and that branch contains commits that revert previous PRs, in order to make it build.
b) There's still some work to be done in the way we're handling GL on the C++ side of things.

note: This PR integrates also my previous PR with Unicode fixes so we probably won't want to merge until that one has been merged first anyway.

### CODE REVIEW

**Comment by Lord-Kamina** [comment]
Can you guys take this one for a spin? I think I'm done with this (Ideally unicode PR should be merged first though since this one also has the changes from that one)

Report with suggestions/questions/bugs and/or comments on performance, please.

**Comment by nieknooijens** [comment]
I think I have to take my time for this 36-commit-long monster. can't review it at work in my break. :worried:

**Comment by Lord-Kamina** [comment]
Sure, just saying it's not really a WIP anymore.

**Comment by nieknooijens** [comment]
ok I have 2 complaints.

first: add set(PROJECT_VERSION 1.0) to ced/CmakeList.txt
cmake was complaining

second.. oh well:
schermafdruk van 2018-08-21 20-27-39

I have no image at all... the game works, I hear music and I can push buttons to control it, but no image...

**Comment by Lord-Kamina** [comment]
The "ced" issue is technically an issue with the unicode PR but ok. 

Re: the image issue, I'll take a look and make sure I didn't omit any commits or anything. What GPU/Distro/Version you on?

**Comment by nieknooijens** [comment]
schermafdruk van 2018-08-21 20-34-27

**Comment by Lord-Kamina** [comment]
Please add the config.xml also.

Finally, can you play around with manually editing the stereo3d settings to different values (on fullscreen ideally) and see if there's any change?

**Comment by nieknooijens** [comment]
[config.xml.txt](https://github.com/performous/performous/files/2307592/config.xml.txt)

made it txt to make github accept it

**Comment by Lord-Kamina** [comment]
Anything out of the ordinary in debug log?

**Comment by Lord-Kamina** [comment]
I've made a bionic install and reproduced your issue. I see nothing in the log or anywhere. I'll see if I can find the issue.

**Comment by nieknooijens** [comment]
nothing in the log, other than "you have UTF-8 issues with pokémon songs" which has been true for a long time now...

**Comment by Lord-Kamina** [comment]
I have pinpointed the issue, it's caused by libglm 0.9.9.0 (windows and mac were on 0.9.8.x); I don't think it's a bug in the library but rather the update removed deprecated or wrong things that might be highlighting issues on our code.

**Comment by Lord-Kamina** [comment]
@nieknooijens Fixed now.

In case you're interested, here's the cause: In our old code with the in-house glm, we initialized all matices as identity; when I changed to libglm I realized there's no identity function, since they were default-constructed to identity. libglm 0.9.9.0 removed the default constructor, which resulted in this error. To fix it, I had to expressly initialize vector types to 0.0f and matrix types to 1.0f, which is how an identity matrix is built according to the GLSL specification.

**Comment by nieknooijens** [comment]
so basically we were using uninitialized memory?
I'll try checking it out tonight and otherwise tomorrow.

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "ced"]
+	path = ced
+	url = https://Lord-Kamina@github.com/performous/compact_enc_det
```
Comment (nieknooijens) [inline .gitmodules]: make this link universal please, when I wanted to commit a fix it started to ask for your password.

File: data/config/schema.xml [hunk 2]
```diff
@@ -94,6 +94,10 @@ to save the current settings to XML.
 		<short>Webserver access</short>
 		<long>Define if you want to control performous remotely with web interface (requires restart).</long>
 	</entry>
+	<entry name="game/case-sorting" type="bool" value="false">
+		<short>Case-sensitive sorting.</short>
+		<long>Make a difference between uppercase and lowercase letters when sorting.</long>
```
Comment (nieknooijens) [inline data/config/schema.xml]: nice one!

File: .gitmodules (hunk 1 above)
Comment (Lord-Kamina) [reply .gitmodules]: Hahahaha hadn't even noticed.

**Comment by Lord-Kamina** [comment]
Will fix the .gitmodules and merge tonight then, unless @Baklap4 wants to weigh-in first?

**Comment by Baklap4** [comment]
I haven't had time to review yet, will do by tonight as i have some time free now 👍

**Comment by Baklap4** [comment]
I think it'd be wise to rebase this upon current master to remove the overlapping commits

**Comment by Lord-Kamina** [comment]
@Baklap4 things I wouldn't wish upon my worst enemy: Solving merge conflicts from a phone. Let's hope I didn't break anything 😂

**Comment by Baklap4** [comment]
Hmm if i'm right you merged master into this branch (resulting in conflicts currently)
If you rebase it (`git checkout master && git pull && git checkout issue/gl_fixes && git rebase -i master`) you'll still have to solve the conflicts, but will remove the commits already made on master disappear from this PR. Resulting in a force push being needed since the local branch and remote branch are diversed. (`git push -f`)

Also this branch still contains the reverted commits (webserver + cache)

**Comment by Lord-Kamina** [comment]
The reverted commits are reverted again. It's a double negation i.e. It just looks ugly. I haven't manually merged, rather just solved conflicts caused by the gitmodules change.

**Comment by Lord-Kamina** [comment]
@Baklap4 I re-merged and rebased stuff to correct some commit messages and remove the reverted and reverted reverted commits. Please make sure I didn't unwittingly break anything before eventual merge?

**Comment by Baklap4** [comment]
I'll try it out tonight to make sure it doesn't break stuff ;) Thanks for rebasing!

**Comment by Lord-Kamina** [comment]
> [32a37fc](https://github.com/performous/performous/commit/32a37fc7949e0c6d478adb5df765a3c9fba78e74) is doing some weird stuff with Webserver code

Kinda sorta.  It's because I originally began work on this branch on the parallel win-master branch (because the motivation was to fix the Windows rendering glitches) where I had reverted the switch to CppRest in order to be able to build.

**Comment by dos1** [comment]
OK, after rechecking, seems this branch works for me and fixes the segfault from #401 :)

However, window resize is now broken - the content inside always stays at the same, initial size.

**Comment by Lord-Kamina** [comment]
> OK, after rechecking, seems this branch works for me and fixes the segfault from #401 :)
> 
> However, window resize is now broken - the content inside always stays at the same, initial size.

Can you try tracing the resize issue to see whether I made a mistake in my code (quite possible)

**Comment by Lord-Kamina** [comment]
@baklap4 can you see whether you have any issues resizing? I'm fairly sure I tested it without issue but I DID mess with resizing code a bit so I might have missed something.

**Comment by Baklap4** [comment]
i can resize without problems however i can resize to a smaller window as performous can show.. resulting in images like this: 
image

maybe let's set a min width and height to which max may be downsized?

**Comment by Lord-Kamina** [comment]
What happens if after resizing, you toggle fullscreen?

Most likely I made a mistake. Will take a look tonight/tomorrow

**Comment by Baklap4** [comment]
If you resized your window way smaller and go full screen it goes full screen and looking alright. If you disable fullscreen again it'll reset your window to contain the default resolution. (which is good)

**Comment by Lord-Kamina** [comment]
Resizing should be fixed now.

**Comment by Baklap4** [comment]
This branch currently has conflicting files..

**Comment by Lord-Kamina** [comment]
I'll check tonight since github seemingly won't allow me to solve the conflicts via its UI

**Comment by Lord-Kamina** [comment]
Just noticed I'd broken it with the merge the other day, I must have been very tired because I'd resolved the conflict as SWFormat and AVResample 😂

**Comment by Lord-Kamina** [comment]
I have re-rebased this monster; theoretically the cpprestsdk reverts should be really gone and not just disguised now.

**Pull request merged**